### PR TITLE
feat: implement Watch with label selectors

### PR DESCRIPTION
**Pull request by smira**
COSI state can now watch list of resources with filters (based on label
selectors): events are filtered based on the label selectors, and if
change to the labels makes resource match/stop matching the query,
synthetic create/destroy event is pushed.

As part of the change, COSI now also publishes old resource state in
update events.

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

### CODE REVIEW

File: pkg/state/options.go [hunk 1]
```diff
@@ -183,3 +184,41 @@ func WithKindTailEvents(n int) WatchKindOption {
 		opts.TailEvents = n
 	}
 }
+
+// WatchWithLabelQuery appends a label query to the list options.
+func WatchWithLabelQuery(query resource.LabelQuery) WatchKindOption {
```
Comment (Unix4ever) [inline pkg/state/options.go:189]: I wish we could reuse the same methods for list and watch as now the API will look a bit unobvious.
There's `WithLabelQuery` which only works with the list requests. 
Or we should rename `WithLabelQuery` to `ListWithLabelQuery`.

Or another idea is to have subfolder `watch` to make it be `watch.WithLabelQuery`.

File: pkg/state/protobuf/client/client.go [hunk 2]
```diff
@@ -384,6 +375,20 @@ func watchAdapter(ctx context.Context, cli v1alpha1.State_WatchClient, ch chan<-
 			return
 		}
 
+		if msg.Event.Old != nil {
+			unmarshaled, err = protobuf.Unmarshal(msg.Event.Old)
+			if err != nil {
+				// no way to signal error here?
```
Comment (Unix4ever) [inline pkg/state/protobuf/client/client.go:381]: maybe COSI resource should also have the error field in the metadata ;)

File: pkg/state/options.go (hunk 1 above)
Comment (smira) [reply pkg/state/options.go:189]: it's still possible to do that bypassing `With` methods, I will think about better option tomorrow... I don't like it either

**Review by Unix4ever — APPROVED** [review]

**Pull request merged**